frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseRenameAction.java

```diff
@@ -0,0 +1,105 @@
+
+import java.util.Collections;
+
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.project.Project;
+import git4idea.branch.GitBranchOperationType;
+import git4idea.branch.GitBrancher;
+import git4idea.branch.GitNewBranchDialog;
+import git4idea.i18n.GitBundle;
+import git4idea.repo.GitRepository;
+import lombok.experimental.ExtensionMethod;
+import lombok.val;
+import org.checkerframework.checker.guieffect.qual.UIEffect;
+
+import com.virtuslab.binding.RuntimeBinding;
+import com.virtuslab.branchlayout.api.BranchLayout;
+import com.virtuslab.branchlayout.api.readwrite.IBranchLayoutWriter;
+import com.virtuslab.gitmachete.frontend.actions.backgroundables.RenameBackgroundable;
+import com.virtuslab.gitmachete.frontend.actions.expectedkeys.IExpectsKeyGitMacheteRepository;
+import com.virtuslab.gitmachete.frontend.resourcebundles.GitMacheteBundle;
+import com.virtuslab.qual.async.ContinuesInBackground;
+import com.virtuslab.qual.guieffect.IgnoreUIThreadUnsafeCalls;
+
+@ExtensionMethod(GitMacheteBundle.class)
+public abstract class BaseRenameAction extends BaseGitMacheteRepositoryReadyAction
+    implements
+      IBranchNameProvider,
+      IExpectsKeyGitMacheteRepository {
+
+  @Override
+  @UIEffect
+  protected void onUpdate(AnActionEvent anActionEvent) {
+    super.onUpdate(anActionEvent);
+
+    val presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+
+    val branchName = getNameOfBranchUnderAction(anActionEvent);
+    val branch = branchName != null
+        ? getManagedBranchByName(anActionEvent, branchName)
+        : null;
+
+    if (branch == null) {
+      presentation.setEnabled(false);
+      presentation.setDescription(getNonHtmlString("action.GitMachete.description.disabled.undefined.machete-branch")
+          .fmt("Rename", getQuotedStringOrCurrent(branchName)));
+    } else {
+      presentation.setDescription(
+          getNonHtmlString("action.GitMachete.BaseRenameAction.description").fmt(branch.getName()));
+    }
+  }
+
+  @Override
+  @ContinuesInBackground
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    val project = getProject(anActionEvent);
+    val gitRepository = getSelectedGitRepository(anActionEvent);
+    val currentBranchName = getNameOfBranchUnderAction(anActionEvent);
+    val branchLayout = getBranchLayout(anActionEvent);
+    val branchLayoutWriter = RuntimeBinding.instantiateSoleImplementingClass(IBranchLayoutWriter.class);
+
+    if (gitRepository == null || currentBranchName == null || branchLayout == null) {
+      return;
+    }
+
+    rename(project, gitRepository, currentBranchName, branchLayout, branchLayoutWriter);
+  }
+
+  @ContinuesInBackground
+  @UIEffect
+  @IgnoreUIThreadUnsafeCalls({"git4idea.branch.GitBrancher$renameBranch(java.lang.String, java.lang.String, java.util.List)"})
```

Comment:
Hmm I'd even add `GitBrancher$renameBranch` (btw, why `$` and not `.` 🤔) just queues a background task, I'd consider it safe for running from UI thread. Add to `com.virtuslab.archunit.UIThreadUnsafeMethodInvocationsTestSuite#whitelistedMethodFullNames_git4idea`

Reply:
Resolved? 🤔